feat(logging): add logging setup

ethanfrogers opened the pull request:
add a library for configuring logging in a common way, including remote
logging which supports armory's troubleshooting platform.

ethanfrogers committed:
feat(logging): add logging setup

add a library for configuring logging in a common way, including remote
logging which supports armory's troubleshooting platform.

ethanfrogers commented:
This PR is essentially a mockup of how we might provide a library for a common way to configure logging for services written in Go. I'm totally open to suggestions on UX and what to improve.

The reason for implementing this in a shared library is that it enabled us to provide a common interface for configuring logging in our Go based services, much like the JVM services use logback. For users running these services, this reduces the amount of knowledge they need to configure logging across all services.

In practice, I'd expect this library to be used like this:
```
import "github.com/armory-io/go-spec/logging"

type MyAppConfig struct {
  // injects logging config into the applications main config file
  Logging logging.Config `json:"logging" yaml:"logging"`
}

config, _ := ImaginaryParseConfigMethod()
// other initialization code omitted for brevity

// for applications using logrus and it's package global logger
logging.ConfigureLogrus(log.StandardLogger(), config.Logger)
```

By default, the logger will remain untouched, leaving all of logrus' defaults untouched. This library also supports configuring JSON logging which can be enabled like this:

```
logging:
  format:
    json: {} <-- open to ideas for making this less wonky
```

I'd also like to support configuration for how certain fields should show up in the output. For example, by default, the JSON formatter outputs messages under the `msg` key. If you'd like it to be `message`, you could configure it like so. 

```
logging:
  format:
    json:
     fields:
       msg: message
       timestamp: "@timestamp"
```

This support use cases where customers already have logging filters(?) setup for specific formats and enables them to get closer to their desired format (i.e. to match the format coming from the JVM services).

ethanfrogers commented:
Chatted with @dogonthehorizon and he recommended a config format like
```
logging:
  format: json
  fields:
    msg: message
    timestamp: "@timetamp"
```

I'm going to attempt to implement this and come back for more feedback.

ethanfrogers committed:
chore(logging): tests and structure

improve the config structure for the logging package and add tests to
ensure proper configuration of various components like JSON formatter
remote logging

ethanfrogers commented:
@dogonthehorizon I started to work on the above but decided to go a different direction. Here's what I settled on config-wise (explanation below):

To enable JSON formatting, 
```
logging:
  json:
    enabled: true
    fields:
      key1: value1
```

When I started implementing support for polymorphic types I realized I'd have to be more prescriptive about how the upstream config was parsed i.e. I'd have to know you were deserializing YAML or JSON using a specific library (most YAML libs don't have a struct similar to `json.RawPayload`). I opted for something more verbose to keep the library useful across various methods of parsing the configuration. Over time, if we find that we want something different, we can change it.

ethanfrogers committed:
chore(example): add usage example

aelath commented:
I think this is pretty great. I don't have strong opinions about how the configuration works.

ncknt commented on logging/configure.go:
```diff
@@ -0,0 +1,105 @@
+
+}
+
+func ConfigureLogrus(l *logrus.Logger, config Config) error {
```
Could we rename so the implementation is not attached to a specific library if we want to swap later on? Would it be worth providing our own interface for `logrus.Logger` that new work would be using?

Having a `NoopLogger` would be nice for testing as well. I can't remember if logrus had any out of the box.

ethanfrogers replied:
@ncknt - Are you suggesting that this library defines the underlying log library instead of the consumer? I.e. I let this library dictate which one I'm using instead of choosing it for myself?

ethanfrogers replied:
@dogonthehorizon - any thoughts on that, assuming I'm interpreting correctly.

ncknt replied:
Yes. The reasoning is to have something similar to slf4j and use a facade. It could be as simple as a `Logger` interface. If we find a better library than logrus, it would be trivial to swap implementation.

ncknt commented on logging/configure.go:
```diff
@@ -0,0 +1,105 @@
+
+type Config struct {
+	Remote RemoteLoggingConfig `json:"remote" yaml:"remote"`
+	JSON   FormatJson          `json:"json" yaml:"json"`
```
Can we also customize non JSON output? I don't know if it's useful but many logging libraries allow you to have different "appenders" with different format. e.g. send to console in human readable format, send JSON to remote location, etc. I have not had the need but perhaps we could either use the same concept of appenders. e.g.

```
logging:
  remote:
     ...
  appenders:
    default:
      destination: stdout
      level: DEBUG
      format: json
      fields: [...]
    myfile:
      destination: /var/log/myservice.log
      format: text
      fields: [...]
```
The use of `destination` is definitely more limited in the container world. `level` seems required though.

ncknt commented on logging/configure.go:
```diff
@@ -0,0 +1,105 @@
+
+	formatter, err := formatters.NewHttpLogFormatter(hostname, config.CustomerID, config.Version)
+	if err != nil {
+		return fmt.Errorf("failed to instantiate remote log formatter: %s", err.Error())
```
Wrap the error with `fmt.Errorf("failed to ....: %w", err)`? I think we're all past Go 1.13.

ethanfrogers replied:
Ha, I think you're right. This code is from a few years ago. Just copy/pasted over.

ncknt commented on logging/configure.go:
```diff
@@ -0,0 +1,105 @@
+		return fmt.Errorf("failed to instantiate remote log formatter: %s", err.Error())
+	}
+
+	if config.Endpoint == "" {
```
Should we default to something? `logging.armory.io`?

ethanfrogers committed:
fix(error): fix error formatting

ethanfrogers committed:
feat(interface): define a logging interface

defines a LeveledLogger interface and a couple of example
implementations for Zap and Logrus as experiments to make sure we can
swap implementations.

ethanfrogers commented:
@ncknt @dogonthehorizon  - I'm still thinking about the appender stuff but I took a stab at defining a `LeveledLogger` interface using both Logrus and Zap as an example (adapters included). I borrowed most of the interface from Logrus and it ended up looking like this:

```
type LeveledLogger interface {
  WithField(key string, value interface{}) LeveledLogger
  WithFields(fields map[string]interface{}) LeveledLogger

  Debugf(format string, args ...interface{})
  Infof(format string, args ...interface{})
  Warnf(format string, args ...interface{})
  Errorf(format string, args ...interface{})
  Fatalf(format string, args ...interface{})
  Panicf(format string, args ...interface{})

  Debug(args ...interface{})
  Info(args ...interface{})
  Warn(args ...interface{})
  Error(args ...interface{})
  Fatal(args ...interface{})
  Panic(args ...interface{})
}
```

Based on applications I've written, I this satisfies my needs. `WithField/WithFields` gives us the ability to make contextual loggers. We have all of the most commonly used log levels (with the exception of `Trace` which I haven't seen widely used). Both Zap and Logrus actually satisfy 85% of the interface natively but we need to use adapters in order to satisfy the `WithField/WithFields` method in a common way. Seems reasonable to me. Would love to hear your thoughts. I've updated the example program to reflect my changes as well.

ethanfrogers commented:
@ncknt - If you're happy with the interface, I'm considering merging this now and coming back to the appender bits. I'd like to make progress unblocking one of our customers instead of prolonging this effort and letting it block that story.
